File: pharmion-backend/Pharmion.Services/Services/PrescriptionService.cs
using MapsterMapper;
using Microsoft.EntityFrameworkCore;
using Pharmion.Model.Enums;
using Pharmion.Model.Exceptions;
using Pharmion.Model.Requests;
using Pharmion.Model.Responses;
using Pharmion.Model.SearchObjects;
using Pharmion.Services.Database;
using Pharmion.Services.Database.Entities;
using Pharmion.Services.Interfaces;

namespace Pharmion.Services.Services
{
    public class PrescriptionService
        : BaseCRUDService<PrescriptionResponse, PrescriptionSearchObject,
                          Prescription, PrescriptionUpsertRequest, PrescriptionUpsertRequest>,
          IPrescriptionService
    {
        public PrescriptionService(PharmionDbContext context, IMapper mapper)
            : base(context, mapper) { }

        public override async Task<PagedResult<PrescriptionResponse>> GetAsync(PrescriptionSearchObject search)
        {
            var query = _context.Prescriptions
                .Include(p => p.Patient)
                .Include(p => p.CreatedByPharmacist)
                .Include(p => p.Items)
                    .ThenInclude(i => i.Product)
                .AsQueryable();

            query = ApplyFilter(query, search);

            int? totalCount = null;
            if (search.IncludeTotalCount)
                totalCount = await query.CountAsync();

            query = query.OrderByDescending(p => p.IssuedAt);

            if (!search.RetrieveAll && search.Page.HasValue && search.PageSize.HasValue)
                query = query.Skip(search.Page.Value * search.PageSize.Value)
                             .Take(search.PageSize.Value);

            var items = await query.ToListAsync();

            return new PagedResult<PrescriptionResponse>
            {
                Items = items.Select(MapToResponse).ToList(),
                TotalCount = totalCount
            };
        }

        public override async Task<PrescriptionResponse?> GetByIdAsync(int id)
        {
            var p = await _context.Prescriptions
                .Include(p => p.Patient)
                .Include(p => p.CreatedByPharmacist)
                .Include(p => p.Items)
                    .ThenInclude(i => i.Product)
                .FirstOrDefaultAsync(p => p.Id == id);

            return p == null ? throw new NotFoundException($"Prescription with ID {id} does not exist") : MapToResponse(p);
        }

        protected override IQueryable<Prescription> ApplyFilter(
            IQueryable<Prescription> query, PrescriptionSearchObject search)
        {
            if (search.PatientId.HasValue)
                query = query.Where(p => p.PatientId == search.PatientId.Value);

            if (search.CreatedByPharmacistId.HasValue)
                query = query.Where(p => p.CreatedByPharmacistId == search.CreatedByPharmacistId.Value);

            if (!string.IsNullOrEmpty(search.DoctorName))
                query = query.Where(p => p.DoctorName.Contains(search.DoctorName));

            if (search.Status.HasValue)
                query = query.Where(p => p.Status == search.Status.Value);

            if (search.IssuedFrom.HasValue)
                query = query.Where(p => p.IssuedAt >= search.IssuedFrom.Value);

            if (search.IssuedTo.HasValue)
                query = query.Where(p => p.IssuedAt <= search.IssuedTo.Value);

            return query;
        }

        protected override async Task BeforeInsert(Prescription entity, PrescriptionUpsertRequest request)
        {
            var patient = await _context.Patients.FindAsync(request.PatientId)
                ?? throw new UserException("Pacijent nije pronađen.");

           
            entity.IssuedAt = DateTime.UtcNow;
            entity.Status = PrescriptionStatus.Active;

            entity.Items = request.Items.Select(i => new PrescriptionItem
            {
                ProductId = i.ProductId,
                Dosage = i.Dosage,
                QuantityPerPeriod = i.QuantityPerPeriod,
                PeriodDays = i.PeriodDays,
                Repeats = i.Repeats,
                RepeatsUsed = 0,
                TherapyType = i.TherapyType,
                NextEligibleDispenseAt = DateTime.UtcNow 
            }).ToList();
        }

        protected override async Task BeforeUpdate(Prescription entity, PrescriptionUpsertRequest request)
        {
            if (entity.Status != PrescriptionStatus.Active)
                throw new UserException("Nije moguće izmijeniti recept koji nije aktivan.");

            
            var oldItems = _context.PrescriptionItems.Where(i => i.PrescriptionId == entity.Id);
            _context.PrescriptionItems.RemoveRange(oldItems);

            entity.Items = request.Items.Select(i => new PrescriptionItem
            {
                ProductId = i.ProductId,
                Dosage = i.Dosage,
                QuantityPerPeriod = i.QuantityPerPeriod,
                PeriodDays = i.PeriodDays,
                Repeats = i.Repeats,
                RepeatsUsed = 0,
                TherapyType = i.TherapyType,
                NextEligibleDispenseAt = DateTime.UtcNow
            }).ToList();
        }

        protected override void MapUpdateToEntity(Prescription entity, PrescriptionUpsertRequest request)
        {
            entity.PatientId = request.PatientId;
            entity.DoctorName = request.DoctorName;
            entity.Facility = request.Facility;
            entity.ValidFrom = request.ValidFrom;
            entity.ValidTo = request.ValidTo;
            entity.Notes = request.Notes;
        }

        public async Task<PrescriptionResponse> CreateAsync(PrescriptionUpsertRequest request, int pharmacistId)
        {
            var patient = await _context.Patients.FindAsync(request.PatientId)
                ?? throw new UserException("Pacijent nije pronađen.");

            var prescription = new Prescription
            {
                PatientId = request.PatientId,
                CreatedByPharmacistId = pharmacistId, 
                DoctorName = request.DoctorName,
                Facility = request.Facility,
                IssuedAt = DateTime.UtcNow,
                ValidFrom = request.ValidFrom,
                ValidTo = request.ValidTo,
                Notes = request.Notes,
                Status = PrescriptionStatus.Active,
                Items = request.Items.Select(i => new PrescriptionItem
                {
                    ProductId = i.ProductId,
                    Dosage = i.Dosage,
                    QuantityPerPeriod = i.QuantityPerPeriod,
                    PeriodDays = i.PeriodDays,
                    Repeats = i.Repeats,
                    RepeatsUsed = 0,
                    TherapyType = i.TherapyType,
                    NextEligibleDispenseAt = DateTime.UtcNow
                }).ToList()
            };

            _context.Prescriptions.Add(prescription);
            await _context.SaveChangesAsync();

            return await GetByIdAsync(prescription.Id)
                ?? throw new UserException("Greška pri kreiranju recepta.");
        }

        public async Task CancelAsync(int id)
        {
            var prescription = await _context.Prescriptions.FindAsync(id)
                ?? throw new UserException("Recept nije pronađen.");

            if (prescription.Status == PrescriptionStatus.Cancelled)
                throw new UserException("Recept je već otkazan.");

            prescription.Status = PrescriptionStatus.Cancelled;
            await _context.SaveChangesAsync();
        }

        private PrescriptionResponse MapToResponse(Prescription p) => new()
        {
            Id = p.Id,
            PatientId = p.PatientId,
            PatientName = p.Patient != null
                ? $"{p.Patient.FirstName} {p.Patient.LastName}"
                : string.Empty,
            CreatedByPharmacistId = p.CreatedByPharmacistId,
            PharmacistName = p.CreatedByPharmacist != null
                ? $"{p.CreatedByPharmacist.FirstName} {p.CreatedByPharmacist.LastName}"
                : string.Empty,
            DoctorName = p.DoctorName,
            Facility = p.Facility,
            IssuedAt = p.IssuedAt,
            ValidFrom = p.ValidFrom,
            ValidTo = p.ValidTo,
            Status = p.Status,
            StatusDisplay = p.Status.ToString(),
            Notes = p.Notes,
            Items = p.Items.Select(i => new PrescriptionItemResponse
            {
                Id = i.Id,
                PrescriptionId = i.PrescriptionId,
                ProductId = i.ProductId,
                ProductName = i.Product?.Name ?? string.Empty,
                Dosage = i.Dosage,
                QuantityPerPeriod = i.QuantityPerPeriod,
                PeriodDays = i.PeriodDays,
                Repeats = i.Repeats,
                RepeatsUsed = i.RepeatsUsed,
                TherapyType = i.TherapyType,
                TherapyTypeDisplay = i.TherapyType.ToString(),
                LastDispensedAt = i.LastDispensedAt,
                NextEligibleDispenseAt = i.NextEligibleDispenseAt
            }).ToList()
        };
    }
}

File: pharmion-frontend/pharmion_desktop/lib/services/prescription_service.dart
import 'api_service.dart';

class PrescriptionItemModel {
  final int id;
  final int prescriptionId;
  final int productId;
  final String productName;
  final String dosage;
  final int quantityPerPeriod;
  final int periodDays;
  final int repeats;
  final int repeatsUsed;
  final String therapyType;
  final String therapyTypeDisplay;
  final DateTime? lastDispensedAt;
  final DateTime? nextEligibleDispenseAt;

  const PrescriptionItemModel({
    required this.id,
    required this.prescriptionId,
    required this.productId,
    required this.productName,
    required this.dosage,
    required this.quantityPerPeriod,
    required this.periodDays,
    required this.repeats,
    required this.repeatsUsed,
    required this.therapyType,
    required this.therapyTypeDisplay,
    this.lastDispensedAt,
    this.nextEligibleDispenseAt,
  });

  bool get isExhausted => repeatsUsed >= repeats;
  int get repeatsRemaining => repeats - repeatsUsed;

  factory PrescriptionItemModel.fromJson(Map<String, dynamic> json) {
    return PrescriptionItemModel(
      id: json['id'] as int? ?? 0,
      prescriptionId: json['prescriptionId'] as int? ?? 0,
      productId: json['productId'] as int? ?? 0,
      productName: json['productName'] as String? ?? '',
      dosage: json['dosage'] as String? ?? '',
      quantityPerPeriod: json['quantityPerPeriod'] as int? ?? 0,
      periodDays: json['periodDays'] as int? ?? 0,
      repeats: json['repeats'] as int? ?? 0,
      repeatsUsed: json['repeatsUsed'] as int? ?? 0,
      therapyType: json['therapyType'] as String? ?? '',
      therapyTypeDisplay: json['therapyTypeDisplay'] as String? ?? '',
      lastDispensedAt: json['lastDispensedAt'] != null
          ? DateTime.tryParse(json['lastDispensedAt'])
          : null,
      nextEligibleDispenseAt: json['nextEligibleDispenseAt'] != null
          ? DateTime.tryParse(json['nextEligibleDispenseAt'])
          : null,
    );
  }
}

class PrescriptionModel {
  final int id;
  final int patientId;
  final String patientName;
  final int createdByPharmacistId;
  final String pharmacistName;
  final String doctorName;
  final String? facility;
  final DateTime issuedAt;
  final DateTime? validFrom;
  final DateTime? validTo;
  final String status;
  final String statusDisplay;
  final String? notes;
  final List<PrescriptionItemModel> items;

  const PrescriptionModel({
    required this.id,
    required this.patientId,
    required this.patientName,
    required this.createdByPharmacistId,
    required this.pharmacistName,
    required this.doctorName,
    this.facility,
    required this.issuedAt,
    this.validFrom,
    this.validTo,
    required this.status,
    required this.statusDisplay,
    this.notes,
    required this.items,
  });

  bool get isActive => status == 'Active';
  bool get isExpired =>
      validTo != null && validTo!.isBefore(DateTime.now());

  factory PrescriptionModel.fromJson(Map<String, dynamic> json) {
    return PrescriptionModel(
      id: json['id'] as int? ?? 0,
      patientId: json['patientId'] as int? ?? 0,
      patientName: json['patientName'] as String? ?? '',
      createdByPharmacistId: json['createdByPharmacistId'] as int? ?? 0,
      pharmacistName: json['pharmacistName'] as String? ?? '',
      doctorName: json['doctorName'] as String? ?? '',
      facility: json['facility'] as String?,
      issuedAt: DateTime.tryParse(json['issuedAt'] ?? '') ?? DateTime(2000),
      validFrom: json['validFrom'] != null
          ? DateTime.tryParse(json['validFrom'])
          : null,
      validTo: json['validTo'] != null
          ? DateTime.tryParse(json['validTo'])
          : null,
      status: json['status'] as String? ?? '',
      statusDisplay: json['statusDisplay'] as String? ?? '',
      notes: json['notes'] as String?,
      items: (json['items'] as List<dynamic>?)
              ?.map((i) =>
                  PrescriptionItemModel.fromJson(i as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class PagedPrescriptions {
  final List<PrescriptionModel> items;
  final int totalCount;
  const PagedPrescriptions({required this.items, required this.totalCount});
}

class PrescriptionService {
  static Future<PagedPrescriptions> getPrescriptions({
    int page = 0,
    int pageSize = 10,
    int? patientId,
    String? status,
    String? patientName,
  }) async {
    final params =
        StringBuffer('Prescription?includeTotalCount=true&retrieveAll=false');
    params.write('&page=$page&pageSize=$pageSize');
    if (patientId != null) params.write('&patientId=$patientId');
    if (status != null && status.isNotEmpty) params.write('&status=$status');
    if (patientName != null && patientName.isNotEmpty) 
    params.write('&patientName=$patientName');

    final data =
        await ApiService.get(params.toString()) as Map<String, dynamic>;
    final rawItems = (data['items'] as List?) ?? [];

    return PagedPrescriptions(
      items: rawItems
          .map((p) =>
              PrescriptionModel.fromJson(p as Map<String, dynamic>))
          .toList(),
      totalCount: data['totalCount'] as int? ?? 0,
    );
  }

  static Future<PrescriptionModel> getById(int id) async {
    final data =
        await ApiService.get('Prescription/$id') as Map<String, dynamic>;
    return PrescriptionModel.fromJson(data);
  }

  static Future<PrescriptionModel> create(
      Map<String, dynamic> request) async {
    final data = await ApiService.post('Prescription', request)
        as Map<String, dynamic>;
    return PrescriptionModel.fromJson(data);
  }

  static Future<PrescriptionModel> update(
      int id, Map<String, dynamic> request) async {
    final data = await ApiService.put('Prescription/$id', request)
        as Map<String, dynamic>;
    return PrescriptionModel.fromJson(data);
  }

  static Future<void> delete(int id) async {
    await ApiService.delete('Prescription/$id');
  }

  static Future<void> cancel(int id) async {
  await ApiService.post('Prescription/$id/cancel', {});
  }

  static List<String> get allStatuses => ['Active', 'Completed', 'Cancelled'];

  static String therapyTypeDisplay(String type) {
    switch (type) {
      case 'ChronicMonthly':
        return 'Chronic (Monthly)';
      case 'ChronicQuarterly':
        return 'Chronic (Quarterly)';
      case 'Acute':
        return 'Acute';
      default:
        return type;
    }
  }
}